src/async_fuse/memfs/kv_engine/mod.rs

```diff
@@ -0,0 +1,377 @@
+            KeyType::INum2DirEntry(ref i) => format!("D{i}").into_bytes(),
+            KeyType::Path2INum(ref p) => format!("P{p}").into_bytes(),
+            KeyType::INum2Attr(ref i) => format!("A{i}").into_bytes(),
+        }
+    }
+}
+
+/// The Txn is used to provide support for metadata.
+#[async_trait]
+pub trait MetaTxn {
+    /// Get the value by the key.
+    /// Notice : do not get the same key twice in one transaction.
+    async fn get(&mut self, key: &KeyType) -> DatenLordResult<Option<ValueType>>;
+    /// Set the value by the key.
+    async fn set(&mut self, key: &KeyType, value: &ValueType) -> DatenLordResult<()>;
+    /// Delete the value by the key.
+    async fn delete(&mut self, key: &KeyType) -> DatenLordResult<()>;
+    /// Commit the transaction.
+    /// Only when commit is called, the write operations will be executed.
+    /// If the commit is successful, return true, else return false.
+    async fn commit(&mut self) -> DatenLordResult<bool>;
+}
+
+/// To support different K/V storage engines, we need to a trait to abstract the K/V storage engine.
+#[async_trait]
+pub trait KVEngine: Send + Sync + Debug {
+    /// Create a new transaction.
+    async fn new_meta_txn(&self) -> Box<dyn MetaTxn + Send>;
+}
+
+/// The `etcd`'s transaction impl.
+/// The txn won't do anything until commit is called.
+/// Write operations are buffered until commit is called.
+struct EtcdTxn {
+    /// The etcd client.
+    client: etcd_client::Client,
+    /// The key is the key in bytes, the value is the version of the key.
+    version_map: HashMap<Vec<u8>, usize>,
+    /// Store the write operations in the buffer.
+    buffer: HashMap<Vec<u8>, Option<Vec<u8>>>,
+}
+
+impl EtcdTxn {
+    /// Create a new etcd transaction.
+    fn new(client: etcd_client::Client) -> Self {
+        EtcdTxn {
+            client,
+            version_map: HashMap::new(),
+            buffer: HashMap::new(),
+        }
+    }
+}
+
+#[async_trait]
+impl MetaTxn for EtcdTxn {
+    async fn get(&mut self, key: &KeyType) -> DatenLordResult<Option<ValueType>> {
+        // first check if the key is in buffer (write op)
+        let key = key.get_key();
+        assert!(
+            self.buffer.get(&key).is_none(),
+            "get the key after write in the same transaction"
+        );
+        assert!(
+            self.version_map.get(&key).is_none(),
+            "get the key twice in the same transaction"
+        );
+        // Fetch the value from `etcd`
+        let req = etcd_client::EtcdGetRequest::new(key.clone());
+        let mut resp = self
+            .client
+            .kv()
+            .get(req)
+            .await
+            .with_context(|| format!("failed to get GetResponse from etcd, key={key:?}"))?;
+        let kvs = resp.take_kvs();
+        // panic here because we don't expect to have multiple values for one key
+        assert!(kvs.len() <= 1, "multiple values for one key");
```

Comment:
Return an error for bad response?

Reply:
I added `assert` statements for the sake of convenience during development and debugging.They will not be effective in the release versionThese assert statements describe certain scenarios that are not supported by kv_engine, such as `path2inum(key)`. In these scenarios, the caller expects to receive either none or one value, but not two values. Even if the caller were to handle this situation as an error, they would not be able to fix it, as it is considered undefined behavior.

Reply:
`panic` is comment typo error . I will fix it .